### PR TITLE
[WIP] SIP-bridge configuration for spreedbox.

### PR DESCRIPTION
**Pull request by Ivansss**

### CODE REVIEW

File: js/sipbridge.js [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+	    else {
+	        $("#saveSIPEndpointButton").prop("disabled", true);
+	    }
+	}
+
+	var getSIPConfiguration = function() {
+		/*
+			TODO: Get sip endpoints configuration from spreedbox-setup
+		*/
+		var testResponse = {
```
Comment (fancycode) [inline js/sipbridge.js]: Please implement this some more, including error handling, so only the actual call to load the configuration must be replaced.

Also, you will always get a list of endpoints.

File: js/sipbridge.js [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+		/*
+			TODO: Set new sip endpoints configuration into spreedbox-setup
+		*/
+	}
+
+	var showSIPConfiguration = function(sipConf) {
+		/*
+			TODO: Display sipConf
+		*/
+		$('ul#listOfSIPServers').prepend(
```
Comment (fancycode) [inline js/sipbridge.js]: See above, you will always get multiple endpoints.

File: templates/settings-admin.php [hunk 3]
```diff
@@ -5,3 +5,36 @@
 		<a class="button" href="/spreedbox-setup/" target="_blank" >Open Spreedbox setup</a>
 	</p>
 </div>
+
+<div class="section" id="sipBridgeSection">
+	<h2>SIP Bridge</h2>
```
Comment (fancycode) [inline templates/settings-admin.php]: Translation here and below.

File: js/sipbridge.js [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    	$('#hostnameInput, #usernameInput, #passwordInput, #roomnameInput').on('input', function() {
+    		validate();
+    	});
+
+    	getSIPConfiguration();
+
+	});
+
+	var validate = function() {
+		if ($('#hostnameInput').val().length > 0 &&
```
Comment (fancycode) [inline js/sipbridge.js]: As discussed, you should also verify that the same hostname/username combination is only used once, and that the room name also is not used for some other endpoint.

Also, the user should see a reason (i.e. message) why the validation failed.

File: js/sipbridge.js [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+			"password":"1234",
+			"room":"sipRoom"
+		};
+
+		showSIPConfiguration(testResponse);
+	}
+
+	var setSIPConfiguration = function(sipConf) {
+		/*
+			TODO: Set new sip endpoints configuration into spreedbox-setup
```
Comment (fancycode) [inline js/sipbridge.js]: Please implement this some more and prepare for handling error responses from the server. Everything should be ready so only the actual call to save the settings must to be exchanged.

**Comment by Ivansss** [comment]
@fancycode I have commit some changes based on your review notes.

**Comment by Ivansss** [comment]
@fancycode Fixed broken notifications. Please, review 😊

File: js/sipbridge.js [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+			validServerUsernameCombination = isServerUsernameCombinationValid(hostname + username);
+
+		if (validRoomName && validServerUsernameCombination){
+			if (hostname.length > 0 &&
+	        	username.length > 0 &&
+	        	password.length > 0 &&
+	        	roomname.length > 0) {
+	        		$("#saveSIPEndpointButton").prop("disabled", false);
+	        		return;
+	    	}
```
Comment (fancycode) [inline js/sipbridge.js]: Indent

File: js/sipbridge.js [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+			$("#setupSIPEndpoint").removeClass('hidden');
+		});
+
+		$("#cancelSIPEndpointButton").on('click', function() {
+			$('#addSIPEndpoint').toggleClass('hidden');
+			$("#setupSIPEndpoint").toggleClass('hidden');
+		});
+
+		$("#saveSIPEndpointButton").on('click', function() {
+			var	sipEndpoint = {
```
Comment (fancycode) [inline js/sipbridge.js]: Indent here and the lines below.

File: templates/settings-admin.php [hunk 8]
```diff
@@ -1,7 +1,50 @@
+<?php
+/** @var \OCP\IL10N $l */
+style('spreedbox', 'sipbridge');
+script('spreedbox', 'sipbridge');
+?>
+
 <div class="section" id="spreedboxSetupSection">
-	<h2>Spreedbox setup</h2>
-	<em>Spreedbox setup provides configuration for the Spreedbox hardware and network.</em>
+	<h2><?php p($l->t('Spreedbox setup')) ?></h2>
```
Comment (fancycode) [inline templates/settings-admin.php]: IMHO these changes should be in a separate MR as they are unrelated to the SIP bridge configuration.

File: js/sipbridge.js [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+
+			$('#addSIPEndpoint').toggleClass('hidden');
+			$("#setupSIPEndpoint").toggleClass('hidden');
+		});
+
+		$( "#listOfSIPServers" ).on('click', 'li > .icon-delete', function() {
+			var $this = $(this).parent(),
+				endpoint = $this.data("endpoint");
+
+			currentSIPConfiguration.splice(currentSIPConfiguration.indexOf(endpoint), 1);
```
Comment (fancycode) [inline js/sipbridge.js]: In case of an error while saving the new settings, the list of endpoints will not be reset.

File: js/sipbridge.js [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+			TODO: Set new sip endpoints configuration into spreedbox-setup
+		*/
+		var url = "/spreedbox-setup/api/v1/sip-brige";
+
+		$.ajax({
+			type:'POST',
+			url:url,
+			data:sipConf,
+			beforeSend: function (xhr) {
+    			xhr.setRequestHeader ("Authorization", authToken.token_type + ' ' + authToken.access_token_raw);
```
Comment (fancycode) [inline js/sipbridge.js]: Indent.

File: js/sipbridge.js [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+		/*
+			TODO: Get sip endpoints configuration from spreedbox-setup
+		*/
+		var url = "/spreedbox-setup/api/v1/sip-brige";
+
+		$.ajax({
+			type:'GET',
+			url:url,
+			beforeSend: function (xhr) {
+    			xhr.setRequestHeader ("Authorization", authToken.token_type + ' ' + authToken.access_token_raw);
```
Comment (fancycode) [inline js/sipbridge.js]: Indent.

File: js/sipbridge.js [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+			},
+		}).success(function (response) {
+				console.log("Success requesting SIP conf:", response);
+				currentSIPConfiguration = response.result;
+				showSIPConfiguration(currentSIPConfiguration);
+		}).fail(function (response, code) {
+				console.log('Error requesting SIP conf:', code, response);
```
Comment (fancycode) [inline js/sipbridge.js]: This should show an error to the user in the UI.

File: js/sipbridge.js [hunk 13]
```diff
@@ -0,0 +1,225 @@
...
+			data:sipConf,
+			beforeSend: function (xhr) {
+    			xhr.setRequestHeader ("Authorization", authToken.token_type + ' ' + authToken.access_token_raw);
+			},
+		}).success(function (response) {
+				console.log("Success setting SIP conf:", response);
+				currentSIPConfiguration = response.result;
+				showSIPConfiguration(currentSIPConfiguration);
+		}).fail(function (response, code) {
+				console.log('Error setting SIP conf:', code, response);
```
Comment (fancycode) [inline js/sipbridge.js]: This should show an error to the user in the UI.

File: js/sipbridge.js [hunk 14]
```diff
@@ -0,0 +1,225 @@
...
+		/*
+			TODO: Display sipConf
+		*/
+		$('ul#listOfSIPServers').empty();
+
+		$.each(sipConf, function(key, value) {
+			$('ul#listOfSIPServers').prepend(
+				$('<li>')
+					.data("endpoint", value)
+					.html('<b>' + value.hostname + '</b>' +
```
Comment (fancycode) [inline js/sipbridge.js]: Please show at least also the username (maybe even the password). Also the indent in the HTML looks strange (the bullet points are too far left).

File: js/sipbridge.js [hunk 15]
```diff
@@ -0,0 +1,225 @@
...
+					};
+
+			currentSIPConfiguration.push(sipEndpoint);
+			setSIPConfiguration(currentSIPConfiguration);
+
+			$('#addSIPEndpoint').toggleClass('hidden');
+			$("#setupSIPEndpoint").toggleClass('hidden');
+		});
+
+		$( "#listOfSIPServers" ).on('click', 'li > .icon-delete', function() {
```
Comment (fancycode) [inline js/sipbridge.js]: There should be a prompt to the user which must be confirmed ("Do you really want to delete the endpoint XXX?"). Otherwise it's too easy to accidentially delete a configuration item.

**Comment by fancycode** [comment]
@Ivansss I updated the JS to use the correct backend APIs and left a couple of `TODO` markers where error handling should be added.

Please update and also check my various other comments.

**Comment by Ivansss** [comment]
@fancycode Please, review latest changes 😊

**Comment by fancycode** [comment]
This most likely needs a UI overhaul (e.g. to get rid of all the inline css and add some better texts), but apart from that looks good code-wise. I'll give it a try on my box soon and will then merge if everything is working fine.

File: js/sipbridge.js [hunk 16]
```diff
@@ -0,0 +1,235 @@
...
+
+	var showSIPConfiguration = function(sipConf) {
+		$('ul#listOfSIPServers').empty();
+
+		$.each(sipConf, function(key, value) {
+			$('ul#listOfSIPServers').prepend(
+				$('<li>')
+					.data("endpoint", value)
+					.html('<span class="status success"></span> ' +
+						  '<b>' + value.username + '@' + value.hostname + '</b>' +
```
Comment (leonklingele) [inline js/sipbridge.js]: This is vulnerable to XSS. Never reflect un-sanitized user input.

File: js/sipbridge.js [hunk 17]
```diff
@@ -0,0 +1,235 @@
...
+ * This file is licensed under the Affero General Public License version 3 or
+ * later. See the COPYING file.
+ *
+ * @author Ivan <ivan@struktur.de>
+ * @copyright struktur AG 2016
+ */
+
+// This file is loaded in ownCloud context
+
+/* globals $, OC, spreedboxAuth */
```
Comment (leonklingele) [inline js/sipbridge.js]: `spreedboxAuth` is not a global. Probably a copy&paste bug.

File: js/sipbridge.js [hunk 18]
```diff
@@ -0,0 +1,235 @@
...
+			
+		$("#saveSIPEndpointButton").prop("disabled", true);
+	}
+
+	var isRoomNameValid = function(roomName) {
+		var valid = true;
+
+		$.each(currentSIPConfiguration, function(key, value) {
+			if (value.room === roomName) {
+				valid = false;
```
Comment (leonklingele) [inline js/sipbridge.js]: Once we're in here (valid == false) we can abort the loop. I think this can be done by returning false from the function.

File: js/sipbridge.js [hunk 19]
```diff
@@ -0,0 +1,235 @@
...
+			}
+		});
+		
+		(valid) ? $('#roomnameErrorMessage').hide() : $('#roomnameErrorMessage').show()
```
Comment (leonklingele) [inline js/sipbridge.js]: This should be a proper if/else block

File: js/sipbridge.js [hunk 20]
```diff
@@ -0,0 +1,235 @@
...
+
+		return valid;
+	}
+
+	var isServerUsernameCombinationValid = function(serverUserCombi) {
+		var valid = true;
+
+		$.each(currentSIPConfiguration, function(key, value) {
+			if ((value.hostname + value.username) === serverUserCombi) {
+				valid = false;
```
Comment (leonklingele) [inline js/sipbridge.js]: Same here. Abort the loop once valid == false

File: js/sipbridge.js [hunk 21]
```diff
@@ -0,0 +1,235 @@
...
+			}
+		});
+
+		(valid) ? $('#usernameErrorMessage').hide() : $('#usernameErrorMessage').show()
```
Comment (leonklingele) [inline js/sipbridge.js]: Proper if/else block

**Comment by Ivansss** [comment]
@leonklingele Please, take a look to the changes.

File: css/sipbridge.css [hunk 22]
```diff
@@ -0,0 +1,24 @@
...
+	margin-left: 10px;
+}
+
+#loadingView {
+	height: 40px;
+}
+
+#listOfSIPServers .status {
+	margin-right: 10px;
+}
```
Comment (leonklingele) [inline css/sipbridge.css]: Missing EOL

File: js/sipbridge.js [hunk 23]
```diff
@@ -0,0 +1,245 @@
...
+
+	var authToken = null,
+		currentSIPConfiguration = [];
+
+	var SERVICE_URL = "/spreedbox-setup/api/v1/system/service/sipbridge";
+
+	var errorSavingConfiguration = OC.L10N.translate('spreedbox', 'Error saving SIP configuration'),
+	    errorGettingConfiguration = OC.L10N.translate('spreedbox', 'Error retrieving SIP configuration');
+
+	$(document).ready(function() {
```
Comment (leonklingele) [inline js/sipbridge.js]: Where does this function end?

File: js/sipbridge.js [hunk 24]
```diff
@@ -0,0 +1,245 @@
...
+			var modifiedSIPConfiguration = currentSIPConfiguration.slice(0);
+
+			modifiedSIPConfiguration.push(sipEndpoint);
+			setSIPConfiguration(modifiedSIPConfiguration, true);
+			$("#saveSIPEndpointButton").prop("disabled", true);
+		});
+
+		$( "#listOfSIPServers" ).on('click', 'li > .icon-delete', function() {
+			var $this = $(this).parent(),
+			    endpoint = $this.data("endpoint");
```
Comment (leonklingele) [inline js/sipbridge.js]: This line is indented using spaces, although L18 uses tabs

File: js/sipbridge.js [hunk 25]
```diff
@@ -0,0 +1,245 @@
...
+			};
+
+			var modifiedSIPConfiguration = currentSIPConfiguration.slice(0);
+
+			modifiedSIPConfiguration.push(sipEndpoint);
+			setSIPConfiguration(modifiedSIPConfiguration, true);
+			$("#saveSIPEndpointButton").prop("disabled", true);
+		});
+
+		$( "#listOfSIPServers" ).on('click', 'li > .icon-delete', function() {
```
Comment (leonklingele) [inline js/sipbridge.js]: You should stick with either single-, or double-quotes. But that would be just nice-to-have :)

File: js/sipbridge.js [hunk 26]
```diff
@@ -0,0 +1,245 @@
...
+			var $this = $(this).parent(),
+			    endpoint = $this.data("endpoint");
+
+			var modifiedSIPConfiguration = currentSIPConfiguration.slice(0);
+
+			modifiedSIPConfiguration.splice(modifiedSIPConfiguration.indexOf(endpoint), 1);
+			setSIPConfiguration(modifiedSIPConfiguration, false);
+		});
+
+    	$('#hostnameInput, #usernameInput, #passwordInput, #roomnameInput').on('input', function() {
```
Comment (leonklingele) [inline js/sipbridge.js]: This and the next few lines are indented using spaces, should use tabs instead

File: js/sipbridge.js [hunk 27]
```diff
@@ -0,0 +1,245 @@
...
+		    roomname = $('#roomnameInput').val(),
+		    validRoomName = isRoomNameValid(roomname),
+		    validServerUsernameCombination = isServerUsernameCombinationValid(hostname + username);
+
+		if (validRoomName && validServerUsernameCombination){
+			if (hostname.length > 0 &&
+			    username.length > 0 &&
+			    password.length > 0 &&
+			    roomname.length > 0) {
+			    	$("#saveSIPEndpointButton").prop("disabled", false);
```
Comment (leonklingele) [inline js/sipbridge.js]: Lines indented using spaces

File: js/sipbridge.js [hunk 28]
```diff
@@ -0,0 +1,245 @@
...
+		});
+
+		if (valid) {
+			$('#roomnameErrorMessage').hide();
+		} else {
+			$('#roomnameErrorMessage').show();
+		}
+
+		return valid;
+	}
```
Comment (leonklingele) [inline js/sipbridge.js]: Every assignment should end with a semicolon (`;`)

File: js/sipbridge.js [hunk 29]
```diff
@@ -0,0 +1,245 @@
...
+		} else {
+			$('#usernameErrorMessage').show();
+		}
+
+		return valid;
+	}
+
+	var getSIPConfiguration = function() {
+		$.ajax({
+			type:'GET',
```
Comment (leonklingele) [inline js/sipbridge.js]: Missing spaces between the colon and the value

File: js/sipbridge.js [hunk 30]
```diff
@@ -0,0 +1,245 @@
...
+
+	var showSIPConfiguration = function(sipConf) {
+		$('ul#listOfSIPServers').empty();
+
+		$.each(sipConf, function(key, value) {
+			$('ul#listOfSIPServers').prepend(
+				$('<li>')
+					.data("endpoint", value)
+					.html('<span class="status success"></span> ' +
+						  '<b>' + escapeHTML(value.username) + '@' + escapeHTML(value.hostname) + '</b>' +
```
Comment (leonklingele) [inline js/sipbridge.js]: Where does `escapeHTML` come from? Is it a Nextcloud feature?

File: templates/settings-admin.php [hunk 31]
```diff
@@ -1,7 +1,54 @@
...
+		
+		<div id="addSIPEndpoint" style="padding-top: 20px">
+			<input type="button" id="addSIPEndpointButton" value="<?php p($l->t('+ Add SIP endpoint')); ?>">
+		</div>
+		
+		<div id="setupSIPEndpoint" style="padding-top: 20px" class="hidden">
+			<p>
+				<input type="text" id="hostnameInput" class="long" placeholder="<?php p($l->t('SIP server')); ?>"/>
+			</p>
+			<p style="padding-top: 10px">
```
Comment (leonklingele) [inline templates/settings-admin.php]: The style should move to a common CSS class declaration

File: templates/settings-admin.php [hunk 32]
```diff
@@ -1,7 +1,54 @@
...
+<div class="section" id="sipBridgeSection">
+	<h2><?php p($l->t('SIP Bridge')) ?></h2>
+	<em><?php p($l->t('The SIP bridge allows users to dial into this Spreedbox and automatically get connected to other users (web or phone) that are in the same room.')) ?></em>
+	<div class="icon-loading" id="loadingView" style="padding-top: 20px"></div>
+	<div class="hidden" id="errorView" style="padding-top: 20px">
+		<p><?php p($l->t('There was a problem while getting current SIP bridge configuration. Please, try again.')) ?></p>
+		<input type="button" id="tryAgainButton" value="<?php p($l->t('Try again')); ?>">
+	</div>
+	<div class="hidden" id="sipEndpointsConfiguration">
+		<div id="savedSIPServers" style="padding-top: 20px">
```
Comment (leonklingele) [inline templates/settings-admin.php]: The style should move to a common CSS class declaration

**Comment by fancycode** [comment]
Closing for now, will be reimplemented with support for enabling/disabling the service and improved error handling. Also will use the status API to show information about SIP registration, current number of calls and web participants.
